compiler/z/codegen/OMRMachine.cpp

```diff
@@ -6124,7 +6124,6 @@ OMR::Z::Machine::initialiseRegisterFile()
 
    // Initialize GPRs
    _registerFile[TR::RealRegister::NoReg] = NULL;
-   _registerFile[TR::RealRegister::SpilledReg] = NULL;
```

Comment:
This change scares me a little. I understand that this is technically incorrect as indexing with `TR::RealRegister::SpilledReg` will go out of bounds of the `_registerFile` and will likely point into `_registerAssociations`

https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/z/codegen/OMRMachine.hpp#L184-L190

Now I cannot say for certain the specific field in `_registerAssociations` that gets accessed will be modified, but it is very likely it will actually remain `NULL` as we clear the associations at every extended basic block:

https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/z/codegen/OMRTreeEvaluator.cpp#L11877-L11882

This is because register associations typically don't hardcode the real register. They are typically left as `AssignAny` dependencies. `TR::RealRegister::SpilledReg` maps to a value of 91, and `TR::RealRegister::NumRegisters` maps to 81, which means we will be accessing the 10th element of `_registerAssociations` which is GPR9. I don't think we ever hardcode dependencies on GPR9 even for linkage so it is very unlikely GPR9 will appear on a dependency list, and hence it is very unlikely it will appear on an association.

This means `_registerFile[TR::RealRegister::SpilledReg]` will be very likely to be `NULL`. There are several places in the codegen that check whether this value is indeed `NULL` and act on it. Removing the above will have undefined consequences even though it is clearly incorrect.

Looking over the Power codegen and in the spirit of https://github.com/eclipse/omr/issues/2459 we should map `TR::RealRegister::NumRegisters` to represent the actual size of the enum and zero initialize all of these arrays for values which are not valid.

See the Power implementation:

https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/p/codegen/OMRMachine.cpp#L140-L148

https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/p/codegen/RealRegisterEnum.hpp#L217-L220

Reply:
I had similar concerns when I was removing that line. However I decided to go with the removal because this is technically undefined behaviour. If this causes future defects then those should be fixed as well. I did not see any issues in my testing and so felt this was a safe enough change. I did not do any performance testing to see if this would have an effect there. If that is worth checking then let me know. 

> This means _registerFile[TR::RealRegister::SpilledReg] will be very likely to be NULL. There are several places in the codegen that check whether this value is indeed NULL and act on it. Removing the above will have undefined consequences even though it is clearly incorrect.

Can you point me to some instances where we check if `_registerFile[TR::RealRegister::SpilledReg]` is `NULL`? Before submitting this PR I went through all occurrences of `_registerFile` inside z/codegen and didn't see any instances where we check if `_registerFile[TR::RealRegister::SpilledReg]` is `NULL` (I don't think we ever do a null check on `_registerFile`). However I could have missed it considering there's like 400+ occurrences. 

Regarding being in sync with how the Power implementation does things, we may have some other issues to deal with first. RealRegisterEnum in the Power implementation lists only real registers (that correspond to actual hardware) except for `NoReg` and `SpilledReg`. On our side, there are concepts such as  `TR::RealRegister::EvenOddPair`:
https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/z/codegen/RealRegisterEnum.hpp#L161-L174

If we change `TR::RealRegister::NumRegisters` to map the actual size of the enum then situations like the following will become problematic:

https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/z/codegen/OMRMachine.cpp#L7482-L7507

I'm pretty sure the loop in the above function is only concerned with actual registers in hardware. So trying to access PseudoRealRegisters defined in `RealRegisterEnum.hpp` such as `TR::RealRegister::LegalOddPair` will result in undefined behaviour currently and segmentation faults when we define `_registerFile[TR::RealRegister::LegalOddRegister] = NULL`. 

There are a couple of workarounds to this issue. We can instead bound the loop to the last "real Register" which in our case is `TR::RealRegister::LastHPR`:
https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/z/codegen/RealRegisterEnum.hpp#L156-L161

Instead of LastHPR, we can add a new entry into the enum which holds the number of "RealRegisters" inside RealRegisterEnum.hpp

We can also borrow from the Power implementation: 
https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/p/codegen/OMRMachine.cpp#L1919-L1926

However in our case we would probably need to do `TR::RealRegister::NumRegisters - 14` (or 15?). This is kind of messy IMO. 

We can also add bound checks using the `isRealReg` and `isPseudoRealReg` queries that I have updated in this PR. However this might cause a performance issue because we will have to add these in many different locations.

Reply:
The best long term solution in my opinion is to remove these "PseudoRealRegisters" from the `RealRegisterEnum` and add them into their own new enum: `PseudoRealRegisters`. The first thing this will do is avoid confusion. When I first started looking into this, `TR::RealRegister::EvenOddPair` sounded like a *real* register. However, I learned later that it's in fact just a concept and it will allow the RA to map to a real register eventually. The second thing this will do is make it clear that `_registerFile` is a list of *real* registers. If we define _registerFile's size to be similar to how it is defined in the Power side, then _registerFile will be a list of real registers, and a bunch of NULL entries at the end for PseudoRealRegisters. I don't believe _registerFile is meant to be used this way. For example, we are never supposed to access the `_registerFile[TR::RealRegister::LegalEvenOfPair]` element. Consider how we handle such pseudo registers inside `OMR::Z::Machine::findBestFreeRegister`: 

https://github.com/eclipse/omr/blob/de70faf87f67ffe9cd9571d30e5a378009b85b7c/compiler/z/codegen/OMRMachine.cpp#L3078-L3102

We never actually try and access `_registerFile[TR::RealRegister::LegalEvenOfPair]`, but instead `LegalEvenOfPair` indicates that we need to find an actual legal even register. 

I haven't looked into the other PseudoRegisters, but I suspect they will be similar. Thoughts?

Reply:
I do like the idea of moving these pseudo registers into a different enum (they are not real registers, and some of them are not registers at all so the name is still up in the air). I do support that idea and it is a much cleaner solution than bundling them as we do today. I suspect we placed them there originally because of convenience.

Never the less I think the primary question we need to answer is how to represent dependencies? These "extra" registers are predominantly used to specify dependencies. Unlike virtual registers dependencies can be associated with "AssignAny" or "EvenOddPair" for example. Register dependencies are effectively tuples where the first element is a virtual register and the second element is _x_. What is _x_ if we go down the route you propose?

Currently _x_ is of type `OMR::RealRegister:RegNum` which is the union of all real registers and these pseudo register things.

Reply:
Will this new enum replace the original `RegNum` enum? Or will it be effectively "inheriting" from the RegNum enum?

Reply:
> Will this new enum replace the original RegNum enum? Or will it be effectively "inheriting" from the RegNum enum?

It would be

Reply:
I've restored this line back into the codebase : 3e8ab62a63402bf83fdd30d629348b5ad51c6f97

I'll open up a new issue + PR to address this issue as this problem will need to be solved in a larger context. For now we should be okay to merge with this "bug" since it doesn't break anything. I'll add the link for the new issue here for record keeping as well.